File: forgerock-auth/src/main/java/org/forgerock/android/auth/SecureCookieJar.java
/*
 * Copyright (c) 2019 - 2020 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth;

import android.content.Context;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import lombok.Builder;
import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;

public class SecureCookieJar implements CookieJar {

    private SingleSignOnManager singleSignOnManager;
    private AtomicReference<Set<Cookie>> cacheRef = new AtomicReference<>();
    private final long cacheIntervalMillis;
    private static final ScheduledExecutorService worker =
            Executors.newSingleThreadScheduledExecutor();
    private final CookieMarshaller cookieMarshaller = new CookieMarshaller();

    @Builder
    public SecureCookieJar(Context context, SingleSignOnManager singleSignOnManager, Long cacheIntervalMillis) {
        this.singleSignOnManager = singleSignOnManager == null ?
                Config.getInstance().getSingleSignOnManager() : singleSignOnManager;
        this.cacheIntervalMillis = cacheIntervalMillis == null ?
                context.getResources().getInteger(R.integer.forgerock_cookie_cache) * 1000 : cacheIntervalMillis;
    }

    @NotNull
    @Override
    public synchronized List<Cookie> loadForRequest(@NotNull HttpUrl httpUrl) {

        Set<Cookie> cookies = cacheRef.get();
        if (cookies == null) {
            cookies = new HashSet<>();
            Collection<String> storedCookies = singleSignOnManager.getCookies();
            if (!storedCookies.isEmpty()) {
                Set<String> updatedCookies = new HashSet<>(storedCookies);
                Iterator<String> iterator = updatedCookies.iterator();
                while (iterator.hasNext()) {
                    Cookie cookie = cookieMarshaller.unmarshal(iterator.next());
                    if (cookie != null) {
                        if (!isExpired(cookie)) {
                            cookies.add(cookie);
                        } else {
                            //Remove expired cookies
                            iterator.remove();
                        }
                    } else {
                        //Failed to parse it
                        iterator.remove();
                    }
                }

                // Some cookies are expired, or failed to parse, remove it
                if (storedCookies.size() != updatedCookies.size()) {
                    cache(cookies);
                    singleSignOnManager.persist(updatedCookies);
                }
            }
        }

        return filter(httpUrl, cookies);
    }

    @Override
    public synchronized void saveFromResponse(@NotNull HttpUrl httpUrl, @NotNull List<Cookie> list) {

        Set<Cookie> cookies = new HashSet<>();
        for (String c : singleSignOnManager.getCookies()) {
            Cookie cookie = cookieMarshaller.unmarshal(c);
            //Remove the same stored cookies
            if (cookie != null && !contains(cookie, list)) {
                cookies.add(cookie);
            }
        }

        for (Cookie cookie : list) {
            if (!isExpired(cookie)) {
                cookies.add(cookie);
            }
        }

        cache(cookies);

        Set<String> updatedCookies = new HashSet<>();
        for (Cookie cookie : cookies) {
            updatedCookies.add(cookieMarshaller.marshal(cookie));
        }

        singleSignOnManager.persist(updatedCookies);
    }

    private boolean contains(Cookie input, Collection<Cookie> cookies) {
        for (Cookie cookie : cookies) {
            if (cookie.name().equals(input.name()) &&
                    cookie.domain().equals(input.domain()) &&
                    cookie.path().equals(input.path())) {
                return true;
            }
        }
        return false;
    }

    private List<Cookie> filter(HttpUrl httpUrl, Set<Cookie> cookies) {
        List<Cookie> result = new ArrayList<>();
        for (Cookie cookie : cookies) {
            if (!isExpired(cookie) &&
                    cookie.matches(httpUrl)) {
                result.add(cookie);
            }
        }
        return result;
    }

    private boolean isExpired(Cookie cookie) {
        return cookie.expiresAt() < System.currentTimeMillis();
    }

    private void cache(Set<Cookie> cookies) {
        if (cacheIntervalMillis > 0) {
            cacheRef.set(cookies);
            worker.schedule(() -> cacheRef.set(null), cacheIntervalMillis, TimeUnit.MILLISECONDS);
        }
    }

}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/SecureCookieJar.kt
/*
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth

import android.content.Context
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.internal.toImmutableList
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

class SecureCookieJar(context: Context,
                      singleSignOnManager: SingleSignOnManager?,
                      cacheIntervalMillis: Long?) : CookieJar, OkHttpCookieInterceptor {

    private val singleSignOnManager: SingleSignOnManager
    private val cacheRef = AtomicReference<MutableSet<Cookie>?>()
    private val cacheIntervalMillis: Long
    private val cookieMarshaller = CookieMarshaller()

    init {
        this.singleSignOnManager = singleSignOnManager ?: Config.getInstance().singleSignOnManager
        this.cacheIntervalMillis = cacheIntervalMillis
            ?: (context.resources.getInteger(R.integer.forgerock_cookie_cache) * 1000).toLong()
    }

    @Synchronized
    override fun loadForRequest(url: HttpUrl): List<Cookie> {
        var cookies = cacheRef.get()
        if (cookies == null) {
            cookies = HashSet()
            val storedCookies = singleSignOnManager.cookies
            if (!storedCookies.isEmpty()) {
                val updatedCookies: MutableSet<String> = HashSet(storedCookies)
                val iterator = updatedCookies.iterator()
                while (iterator.hasNext()) {
                    val cookie = cookieMarshaller.unmarshal(iterator.next())
                    if (cookie != null) {
                        if (!isExpired(cookie)) {
                            cookies.add(cookie)
                        } else {
                            //Remove expired cookies
                            iterator.remove()
                        }
                    } else {
                        //Failed to parse it
                        iterator.remove()
                    }
                }

                // Some cookies are expired, or failed to parse, remove it
                if (storedCookies.size != updatedCookies.size) {
                    cache(cookies)
                    persist(updatedCookies)
                }
            }
        }
        return filter(url, cookies)
    }

    @Synchronized
    override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
        val cookiesContainer: MutableSet<Cookie> = HashSet()
        for (c in singleSignOnManager.cookies) {
            val cookie = cookieMarshaller.unmarshal(c)
            //Remove the same stored cookies
            if (cookie != null && !contains(cookie, cookies)) {
                cookiesContainer.add(cookie)
            }
        }
        for (cookie in cookies) {
            if (!isExpired(cookie)) {
                cookiesContainer.add(cookie)
            }
        }
        cache(cookiesContainer)
        val updatedCookies: MutableSet<String> = HashSet()
        for (cookie in cookiesContainer) {
            updatedCookies.add(cookieMarshaller.marshal(cookie))
        }
        persist(updatedCookies)
    }

    private fun contains(input: Cookie, cookies: Collection<Cookie>): Boolean {
        for (cookie in cookies) {
            if (cookie.name == input.name && cookie.domain == input.domain && cookie.path == input.path) {
                return true
            }
        }
        return false
    }

    private fun filter(url: HttpUrl, cookies: Set<Cookie>): List<Cookie> {
        val result = mutableListOf<Cookie>()
        cookies.filter { !isExpired(it) }
            .filter { it.matches(url) }
            .toCollection(result)

        return intercept(result.toImmutableList())
    }

    private fun isExpired(cookie: Cookie): Boolean {
        return cookie.expiresAt < System.currentTimeMillis()
    }

    private fun cache(cookies: MutableSet<Cookie>) {
        if (cacheIntervalMillis > 0) {
            cacheRef.set(cookies)
            worker.schedule({ cacheRef.set(null) }, cacheIntervalMillis, TimeUnit.MILLISECONDS)
        }
    }

    private fun persist(cookies: Collection<String>) {
        singleSignOnManager.persist(cookies)
        FRLifecycle.dispatchCookiesUpdated(cookies)
    }

    class SecureCookieJarBuilder internal constructor() {
        private lateinit var context: Context
        private var singleSignOnManager: SingleSignOnManager? = null
        private var cacheIntervalMillis: Long? = null
        fun context(context: Context): SecureCookieJarBuilder {
            this.context = context
            return this
        }

        fun singleSignOnManager(singleSignOnManager: SingleSignOnManager?): SecureCookieJarBuilder {
            this.singleSignOnManager = singleSignOnManager
            return this
        }

        fun cacheIntervalMillis(cacheIntervalMillis: Long?): SecureCookieJarBuilder {
            this.cacheIntervalMillis = cacheIntervalMillis
            return this
        }

        fun build(): SecureCookieJar {
            return SecureCookieJar(context, singleSignOnManager, cacheIntervalMillis)
        }

        override fun toString(): String {
            return "SecureCookieJar.SecureCookieJarBuilder(context=$context, " +
                    "singleSignOnManager=$singleSignOnManager, cacheIntervalMillis=$cacheIntervalMillis)"
        }
    }

    companion object {
        private val worker = Executors.newSingleThreadScheduledExecutor()

        @JvmStatic
        fun builder(): SecureCookieJarBuilder {
            return SecureCookieJarBuilder()
        }
    }
}
